### PR TITLE
Implement Saga

### PR DESCRIPTION
**Pull request by mingfang**
https://github.com/uber-go/cadence-client/issues/797

### CODE REVIEW

File: workflow/saga.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        activity interface{}
+        args     []interface{}
+    }
+)
+
+func NewSaga(options SagaOptions) Saga {
+    return &sagaImpl{options: &options}
+}
+
+func (s *sagaImpl) AddCompensation(ctx Context, activity interface{}, args ...interface{}) Saga {
```
Comment (mfateev) [inline workflow/saga.go]: I don't think we want to store context here. For example if workflow is cancelled the contexts created from the root context will be cancelled and the compensation code is not going to execute correctly. I propose to add ctx to Compensate call and use context created from it using Workflow.newDisconnectedContext.

File: workflow/saga.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
+package workflow
```
Comment (mfateev) [inline workflow/saga.go]: I would move it into its own package

File: workflow/saga.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+type (
+    SagaOptions struct {
+        ParallelCompensation bool
+        ContinueWithError    bool
+    }
+
+    Saga interface {
```
Comment (mfateev) [inline workflow/saga.go]: I propose an alternative interface that uses ctx to pass saga around:

sagaCtx := saga.newSaga(ctx, options)
...
saga.addCompensation(ctx, ....)
...
saga.Compensate(ctx)

This approach works with calls to downstream functions without need to pass Saga object explicitly as Context is passed everywhere already.

**Comment by vancexu** [comment]
Hi @mingfang , since it's been a while I wonder do you have any plans to address the feedbacks and move forward on this PR?

**Comment by mingfang** [comment]
> Hi @mingfang , since it's been a while I wonder do you have any plans to address the feedbacks and move forward on this PR?

Sorry I don't have time to make the necessary changes. Please feel free to take over this PR or close it.

**Pull request closed without merging**